lib/pengine/unpack.c

```diff
@@ -2657,7 +2668,7 @@ unpack_migrate_to_failure(pe_resource_t *rsc, pe_node_t *node, xmlNode *xml_op,
     target_migrate_from_id = pe__call_id(target_migrate_from);
 
     if ((target_stop == NULL) || (target_stop_id < target_migrate_from_id)) {
-        /* There was no stop on the source, or a stop that happened before a
+        /* There was no stop on the target, or a stop that happened before a
          * migrate_from, so assume the resource is still active on the target
          * (if it is up).
          */
```

Comment:
Given a scenario:

1) rscA fails to migrate from node1 to node2 with a failed migrate_to on node1
2) rscA fails to stop on node2
3) node2 gets fenced and rscA recovers on node1
4) node2 rejoins, and rscA is considered being "multiple-active" and an arguably unnecessary full recovery is issued

I understand the logic here tends to be conservative for safety. Basically as long as there's no explicit stop operation of the resource on the target, the resource is considered being active on the target.

And indeed it's not easy for pacemaker to determine whether a migrate_to failure no longer matters given the situation and the limited amount of context here.

I'm not sure whether it's safe, for example if there's any operations on the "source" node newer than the failed migrate_to,  we could consider migrate_to is outdated?

Otherwise it might be sensible to require users to manually cleanup the resource, which basically tells pacemaker that it can safely forget the historical migrate_to failure?

Reply:
Fencing should be considered just as good as a stop, but currently a fencing notification just cleans up the history of the node being fenced, so a migrate_to failure on another node isn't affected. However we don't want to clean the migrate_to since a stop is still needed there. Unfortunately the CIB doesn't keep track of fencing so the scheduler can't know a fencing was equivalent to a stop.

Maybe when a node is fenced and the controller processes an implied stop for it, the controller could record a fake stop event in the resource history.

Reply:
Good idea. If we go with faking a stop operation, it seems it needs to be added into the transition graph by scheduler so that controller knows. And It doesn't seem to be able to handle the case where the "target" is remote node though since it will get its resource history cleaned up upon rejoining.

But I cannot think of a better and thorough solution so far. I was thinking if we could dare assume a start operation on the "source" node that's newer the failed migrate_to implies that stop of the resource on both nodes has been taken care of by the previous cluster transition, but then as long as the migrate_to failure is still existing in cib status,  there could be again the situation where we could no longer safely assume so...

Reply:
> I was thinking if we could dare assume a start operation on the "source" node that's newer the failed migrate_to implies that stop of the resource on both nodes has been taken care of by the previous cluster transition, but then as long as the migrate_to failure is still existing in cib status, there could be again the situation where we could no longer safely assume so...

Unless, probably for resources that support live migration, we should record specific resource operations separately rather than keep overriding "*_last_0" and "*_last_failure_0" (e.g. keep losing the useful historical info and context).

It reminds me of another scenario that suffers from losing context by overriding "*_last_0"/"*_last_failure_0". It's easier to produce with cluster property batch-limit=1:

- rscA is live-migrating from node1 to node2.

- Before migrate_to on node1 returns, put node2 into standby.

- Transition aborts upon return of successful migrate_to on node1. New transition is going to stop rscA on both nodes and start it on node1.

- While it is stopping on node1, run something that is going to make the transition abort again like:
crm_resource  --resource rscA --ban --node node2

- Transition aborts upon return of successful stop on node1.

Now although there's a successful stop on node1, but cluster no longer knows there was a partial migration and rscA is still potentially active on node2...

In a production environment, a similar situation caused multiple-active and corruption of the shared data (VM images).

Reply:
> Good idea. If we go with faking a stop operation, it seems it needs to be added into the transition graph by scheduler so that controller knows.

The implied stop is already in the graph. The controller would record the result when confirming that implied stop.

> And It doesn't seem to be able to handle the case where the "target" is remote node though since it will get its resource history cleaned up upon rejoining.

That's not a problem because if the stop is cleaned, so is the migrate_to.

> 
> But I cannot think of a better and thorough solution so far. I was thinking if we could dare assume a start operation on the "source" node that's newer the failed migrate_to implies that stop of the resource on both nodes has been taken care of by the previous cluster transition, but then as long as the migrate_to failure is still existing in cib status, there could be again the situation where we could no longer safely assume so if the "*_last_0" gets overridden by something else ...

> 
> Unless, probably for resources that support live migration, we should record specific resource operations separately rather than keep overriding "__last_0" and "__last_failure_0" (e.g. keep losing the useful historical info and context).

Long ago, a longer set of history was kept. There is some demand for doing that again, just so we can show a more detailed resource history on request. The downside would be exploding the status section size and having a way to periodically clean the oldest entries.

> 
> It reminds me of another scenario that suffers from losing context by overriding "__last_0"/"__last_failure_0". It's easier to produce with cluster property batch-limit=1:
> 
>     * rscA is live-migrating from node1 to node2.
> 
>     * Before migrate_to on node1 returns, put node2 into standby.
> 
>     * Transition aborts upon return of successful migrate_to on node1. New transition is going to stop rscA on both nodes and start it on node1.
> 
>     * While it is stopping on node1, run something that is going to make the transition abort again like:
>       crm_resource  --resource rscA --ban --node node2
> 
>     * Transition aborts upon return of successful stop on node1.
> 
> 
> Now although there's a successful stop on node1, but cluster no longer knows there was a partial migration and rscA is still potentially active on node2...
> 
> In a production environment, a similar situation caused multiple-active and corruption of the shared data (VM images).

Ouch, that is a bad one.

Another possibility might be: if a stop result arrives, but its transition has been aborted, and a migrate_from was scheduled for the resource on another node, erase the resource's history on the other node to force a reprobe.